### PR TITLE
Apps: Add Dictionary

### PR DESCRIPTION
**Pull request by rogual**
Hey, first "substantial" contribution here, so, please excuse me if I did anything wrong.

I'm not sure how polished you guys like things to be before they get accepted. It's perhaps a little scrappy, code wise?

Anyway, main points:
* Offline, local, searchable dictionary app
* Search-as-you-type
* Data file has 77,000 words and is 7.1MiB

Caveats:
* You have to build the data file yourself using the script included, it's not committed to the repo. Hope that's okay; don't know how you guys like to handle stuff like this.
* ~App icon is still TextEditor~
* English dictionary only
* WordNet only has nouns, verbs, adjectives, adverbs. So words like 'if', 'they' or 'whoa' aren't in there.

Screenshot 2021-08-22 at 15 46 11

Stuff that could be added in the future:
* User-selectable dictionaries
* Better data than WordNet (it's not bad, honestly, just needs prepositions etc. from somewhere else)
* "Look Up" as a standard action on selected text throughout the system (this is really nice on Mac)
* Domain tagging (e.g. [science], [military], etc)
* Assistant integration

Stuff I need to do before this PR gets merged:
* [x] Display WordNet license appropriately

### CODE REVIEW

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (sppmacd) [inline Userland/Applications/Dictionary/Dictionary.cpp:56]: Use `dbgln` or `outln` instead of `printf` here.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (sppmacd) [inline Userland/Applications/Dictionary/Dictionary.cpp:68]:
```suggestion
StringView Dictionary::word_at_index(int word_index) const
```

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (sppmacd) [inline Userland/Applications/Dictionary/Dictionary.cpp:83]:
```suggestion
        return StringView(c, n);
```

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (sppmacd) [inline Userland/Applications/Dictionary/Dictionary.cpp:148]: Is there any reason to use `snprintf` here instead of just `String`?

**Review by sppmacd — COMMENTED** [review]
Small things I noticed:

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Thanks! Have now replaced all `printf`s throughout.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Cheers, fixed (and throughout)

File: Userland/Applications/Dictionary/CMakeLists.txt
Comment (sppmacd) [inline Userland/Applications/Dictionary/CMakeLists.txt:5]: What is ImageDecoder and RequestServer used for?

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:146]: What I'm trying to do here is basically just make sure I have a block of 16 bytes that I can do fast binary comparisons with all the way down the index until I hit the right word.

So, that's why I reached for `char[]`, and once I had a `char[]` and some stuff to put in it, I instinctively reached for `snprintf` — though, if there's a preferred way to do it, I'd be happy to, of course. You might have to spell it out for me though 😄

File: Userland/Applications/Dictionary/CMakeLists.txt
Comment (rogual) [reply Userland/Applications/Dictionary/CMakeLists.txt:5]: Probably nothing, this is all copied from TextEditor. I'll remove 'em.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (sppmacd) [reply Userland/Applications/Dictionary/Dictionary.cpp:146]: This would probably require to change `m_index` to something more C++-y, this code feels to be more like C. Maybe it would be better to wait for someone more experienced to take a view on it :)

Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:144]: `m_index` is just a pointer into `wordnet.bin`.

The index is literally just every word, padded to 16 chars, one after the other, like this:

```
0000  a...............
0010  banana..........
0020  catamaran.......
  [...]
9980  yodel...........
9990  zebra...........
```

It starts at the top, looks at the first letter of each word until it gets a match, then continues from there looking at the second letter, etc.

It's kinda dumb but it works and is fast. I started with the dumbest possible thing and then just made it cleverer until it was fast enough.

(It would be even faster to store all the first letters together, then the second ones, and so on, but I think I've optimized this enough. It's decently fast.)

**Comment by brapru** [comment]
As per the [contributing guidelines](https://github.com/SerenityOS/serenity/blob/ca2c81251a7d036f586f3d0a32bc35e2b9f93144/CONTRIBUTING.md), any later fix-up commit should be squashed into the initial commit. 

Additionally, 
> Split your changes into separate, atomic commits (i.e. A commit per feature or fix, where the build, tests and the system are all functioning).

There are changes you make to LibGUI in the second commit that should probably be pushed to their own separate commit, or squashed along with the first one.

**Comment by rogual** [comment]
Ah OK, I did see

> Squash your commits when making revisions after a patch review

but I thought it meant just before merging.

**Comment by rogual** [comment]
OK, all rebased into nice commits 🙂

File: Meta/build-dictionaries.py
Comment (ADKaster) [inline Meta/build-dictionaries.py:54]: I'm not super convinced that this use of the WordLite database complies with their license? For instance, I don't see the License reproduced at all in this binary packer for the data within.

https://wordnet.princeton.edu/license-and-commercial-use

> When writing a paper or producing a software application, tool, or interface based on WordNet, it is necessary to properly cite the source. 

https://wordnet.princeton.edu/citing-wordnet

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [reply Userland/Applications/Dictionary/Dictionary.cpp:150]: Might be best left for a subsequent PR, but, instead of inventing your own packed binary format, how hard would it be to extend LibSQL to be able to parse and walk the database in its unmodified SQLite-3 format?

File: Userland/Libraries/LibGUI/ListView.h
Comment (ADKaster) [inline Userland/Libraries/LibGUI/ListView.h:66]: -1 is a poor value for a placeholder that's not set. This should probably use Optional<int> .

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [reply Userland/Applications/Dictionary/Dictionary.cpp:150]: If you went that route, it would let you use LibSQL internally in the database to query it at runtime, and avoid the awkward pointer math and packed binary format you have here.

File: Meta/build-dictionaries.py
Comment (rogual) [reply Meta/build-dictionaries.py:54]: I know, I mentioned it in the PR description. I think the about box would probably be the best place for this.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:150]: I don't know, but probably very. To be honest, I was more interested in writing a dictionary app than working on an SQL implementation.

When you say awkward pointer math, do you mean all pointer math is awkward, or the pointer math in this particular PR is awkward?

File: Userland/Libraries/LibGUI/ListView.h
Comment (rogual) [reply Userland/Libraries/LibGUI/ListView.h:66]: Is this a Serenity-wide rule? I actually quite like -1 as an "invalid" value for an int that won't be negative, just because it's an efficient use of a machine word. I like Optional for things like objects with no "invalid" value.

I do see several uses of -1 like this in Serenity already, so it's probably not too frowned upon?

e.g.
https://github.com/SerenityOS/serenity/blob/a8967388d3b6986dd10fab193c0b1fef287a4e18/Userland/Shell/Builtin.cpp#L111

https://github.com/SerenityOS/serenity/blob/76a18e1172a9389fad540fcfdf81252e237da2de/Userland/Services/WindowServer/WindowSwitcher.cpp#L61

https://github.com/SerenityOS/serenity/blob/631d36fd9821e48b3b3b53ac609e2e50b254d15b/Userland/Games/Hearts/Player.cpp#L77

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:150]: BTW, I don't want to come across like I think my ad-hoc binary thing is better than a good DB layer. Having systems share common parts like that is really nice.

I would, though, probably prefer to defer that to a future PR, as you said.

I hope this can be considered an attempt at a usable, MVP Dictionary app, rather than the best possible Dictionary app.

File: Meta/build-dictionaries.py
Comment (rogual) [reply Meta/build-dictionaries.py:54]: Help menu item opening the license TextEditor might be good for now, though? I've added a checkbox to the description so we don't forget.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [reply Userland/Applications/Dictionary/Dictionary.cpp:150]: Sure, MVP is a good way to approach this project.

I meant pointer math in general, I didn't actually take that close of a look at this file. Just enough to wonder "wait we started with a database, isn't there a database library in the system that could parse it instead of this manually packed format?"

That being said if code looks like C that's a good sign that there's probably some type safer C++-y way to do it hiding in the wings.

File: Meta/build-dictionaries.py
Comment (ADKaster) [reply Meta/build-dictionaries.py:54]: I would say the about box is a good a place as any, yeah.

File: Userland/Libraries/LibGUI/ListView.h
Comment (ADKaster) [reply Userland/Libraries/LibGUI/ListView.h:66]: I'm not sure on the logic here. The way I'm looking at this commit, You either have a cached width or you don't. A cached width of size -1 is not valid.

 It's the general design policy of this project that if a construct is more expressive, and makes it clearer to the reader what's happening, then that should be the default tool you reach for. 

I'm not sure that the efficiency of this particular algorithm will be that much affected by an extra boolean attached to the cached width. Though if you have some performance numbers that show this really hot path (?) can't afford an extra bool in the cache line, then we can worry about the data locality efficiency of the list view class.

Comment (rogual) [reply Userland/Libraries/LibGUI/ListView.h:66]: I don't have any performance numbers for this and this isn't a hot path. I just wanted to gently push back on this suggestion because this use of -1 seems quite common in this project. It sounds like you're still quite keen on Optional here, though, so I'm happy to use it.

Comment (rogual) [reply Userland/Libraries/LibGUI/ListView.h]: Done! 🙂

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:150]: Bit more info on the WordNet DB:
* SQLite isn't its original format, the maintainers distribute it in a custom text format and it's someone else that maintains the SQLite distro.
* The WordNet SQLite .db file is 400Mib (100Mib zipped), and I just get the sense that Serenity likes things to be smaller and tighter than that. Of course, you can cut it down a bit, but at that point you're no longer just parsing the file in its original format.

On C-like code, I tend to agree. If I am doing pointer/byte-level stuff I do at least like to wrap it in a nice C++ interface, which is what the `Dictionary` class is intended as.

It would be possible to be a bit more C++-y even there, though, for sure. For instance, whenever I have an offset into the mapped file, I considered having, instead of `int32_t`, an `Offset<T>` which dereferences to a `T`.

I decided against it in the end  — the cost/benefit ratio didn't feel quite right — but I could do it like that, if you like, if it'd improve this PR's chances 🙂

Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:148]: I've updated Dictionary.cpp to make it more type-safe. This removes almost all of the ugly casting and pointer arithmetic. I think it's a big improvement, so thanks for pushing me in that direction. What do you think?

File: Meta/build-dictionaries.py
Comment (rogual) [reply Meta/build-dictionaries.py:96]: OK, license added. I put it as its own Help menu entry because the About box isn't currently customizable and I didn't fancy trying to make that change as well 🙂

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:29]: Please make this helper return a StringView and use StringView literals

```c++
static StringView part_of_speech_name(char pos)
{
    switch (pos) {
    case 'n':
        return "noun"sv;
    case 'v':
        return "verb"sv;
     ...
}
```

File: Userland/Applications/Dictionary/Dictionary.h
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.h:47]: Conventionally we use `u8`, `u32` etc from AK/Types.h

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:85]: Technically String and StringView are nullable, (is_empty() and is_null() both return true for Strings/StringViews constructed via nullptr). Not particularly partial to either `nullptr` or `""` for the not found case, but if you're going to use `""` might as well use `""sv` to save any remote possibility that the compiler tries to call __builtin_strlen on an empty string literal.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:90]: This construct seems kinda weird, might apply DeMorgan's?

```suggestion
    if (word_index < 0 || word_index >= m_num_words) {
```

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:55]: Is the Dictionary a useful object if it has no data? Returning from the constructor with no words and no bytes is a bit of a :/ moment. I would recommend changing the constructor to take a .. `NonnullRefPtr<MappedFile>`. Then you can make the constructor private and use a static factory function to create one. Either returning an `OwnPtr<Dictionary>`  or a `Result<Dictionary, OsError>` to pass the error from the file mapping creation up to the caller.

You could also make m_dictionary_file a `NonnullRefPtr<MappedFile>` this way.
```c++
Dictionary::Dictionary(NonnullRefPtr<MappedFile> dictionary_file)
   : m_dictionary_file(move(dictionary_file))
{
```

Adopting this design pattern would let your UI layer have some logic to like, throw up a message box and say
> Unable to open default dictionary file "/res/dictionaries/wordnet.bin". Please select a valid dictionary or exit

And throw up a file picker dialog or exit. For example.

It would also let you move the verification of the header validity to the factory function, and let you complain to the user that their database is malformed rather than VERIFY and crash the app.

This pattern would work best if the DictionaryModel was given a reference (or ownership of) a Dictionary object in its constructor, and you did that "dictionary selection" logic in main.cpp.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:54]: Not sure where/if it shows up elsewhere, but please use `nullptr` for the null pointer constant.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:121]:
```suggestion
            string_builder.appendff("  {}. ", sense_number + 1);
```

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:129]: returning builder.string_view() here forces construction of a String from a temporary StringView. should use .build() or .to_string() instead.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:144]: please check query.is_empty() instead of directly comparting with ""

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:152]: What happens here if the query is larger than 16 characters? You'll overrun the stack buffer you have here.

Instead of this copy, you could iterate from 0 to the min(16, query.length()) or some such in the for loop below.

Additionally, the query string is only used as a const reference, so there's no need for the caller to have a String at all. query could be replaced with a StringView, and then the loop below could use query.to_lowercase_string() to iterate.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.cpp:161]: I'd recommend using AK/CharacterTypes.h for this instead of using bitwise math here for clarity.

For example, if you iterate from 0 to min(query.length(), 16) as suggested above, you could do something like
```c++
  auto query_lowercase = query.to_lowercase(); // creates a new String on the heap, fwiw
  const size_t max_letters = min(query_lowercase.length(), 16);
  for (size_t letter_index = 0; letter_index < max_letters; ++letter_index) {
     char query_char = query_chars[letter_index];
     char const* p_item_char = &m_index[first_index][letter_index];
     while (to_ascii_lowercase(*p_item_char) < query_char) {
          ...
```

File: Userland/Applications/Dictionary/Dictionary.h
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.h:65]: primitive types (arithmetic types, pointers, etc) conventionally use designated initializers in the class declaration.

```c++
u8 const* m_dictionary_bytes { nullptr };

int m_num_words { 0 };
int m_max_word_length { 0 };
```

File: Userland/Applications/Dictionary/Dictionary.h
Comment (ADKaster) [inline Userland/Applications/Dictionary/Dictionary.h:23]: The name here suggests that `count_words()` will be performing some work to count how many words there are in the dictionary. Since it's not doing that, I'd rename it to `word_count()` or similar.

File: Userland/Applications/Dictionary/DictionaryModel.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/DictionaryModel.cpp:54]: since you're saving the caller's query here, should `move()` it into place.

```c++
        m_query = move(query);
```

File: Userland/Applications/Dictionary/MainWidget.cpp
Comment (ADKaster) [inline Userland/Applications/Dictionary/MainWidget.cpp:18]: Suggest taking a StringView here so that you don't have to strnlen the query to set it as the text for the search box and the model's query.

**Review by ADKaster — CHANGES_REQUESTED** [review]
I hadn't looked at the style much before you C++-ified it, but it looks pretty close to Serenity style, just a few things that'll likely need to be updated before the maintainers will accept it 👇

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Thanks for the context on the wordnet database! I'm not 100% sold on the Offset<T> wrapper, but it seems to be getting the job done. I'm not sure how C-like it was before b/c I didn't look that closely but it seems in decentish shape at this point to me.  I left a longer review below.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: You're right, this is literally just strnlen. Fixed.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Done.

File: Userland/Applications/Dictionary/Dictionary.h
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.h]: Done.

File: Userland/Applications/Dictionary/Dictionary.h
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.h]: Done.

File: Userland/Applications/Dictionary/Dictionary.h
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.h]: Done.

File: Userland/Applications/Dictionary/MainWidget.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/MainWidget.cpp]: Done.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Done.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Nice. What does ff stand for?

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:128]: Done.

File: Userland/Applications/Dictionary/DictionaryModel.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/DictionaryModel.cpp]: Done.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (ADKaster) [reply Userland/Applications/Dictionary/Dictionary.cpp]: I think it's for `fformat`. Joking aside, it's a wip refactor from the old append which takes a printf style format string.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:145]: The index only contains ASCII A-Z and a-z, so I don't really love using `to_ascii_lowercase` here, which will do a check that each char in the index is between `'A'` and `'Z'` every time.

I knew the unadorned `| 32` could be unclear, but I think the comment solves that. I suppose I could make it a function instead, but it didn't seem worth it?

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: You're absolutely right, that will crash, thanks. Fixed.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: 👍

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Nice, thanks.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp]: Changed to `""sv`.

File: Userland/Applications/Dictionary/Dictionary.cpp
Comment (rogual) [reply Userland/Applications/Dictionary/Dictionary.cpp:39]: The app already does show a similar message if you haven't built the data file. It determines whether or not to do this by seeing whether or not the dictionary is empty.

About verification of header validity, there is none. At this stage I'm thinking of the data file really as part of the app; you build it along with the app, and it maps it at startup. Like, it might as well be in the executable's `.data` section, it's that tightly-coupled.

I think doing proper validation is super great and necessary once we have user-selectable dictionaries and we're loading them from files associated with the app, but for what we have now I honestly think it's okay.

**Comment by ADKaster** [comment]
@rogual, I think you have the core of a good feature here, but I'm not sure that the presentation in your application is going to make the maintainers feel confident enough in the application to merge it. I'd recommend hopping in the discord in the #development or #userland channels to discuss with folks where/how they see the dictionary app/widget fitting into the system.

The application as-is is quite basic. I think that the list-view presentation of the dictionary items is probably the simplest possible way  to display the data, and the UX for someone trying to find words is not particularly great with the default scrollable widget scroll bar.

The database of words itself is a good bit of kit, and if it could be more tightly integrated into, for example, the GUI library as a generic Dictionary backend for integration into things like a SpellChecker widget or similar I think it would gain more traction. The fact that the user can't add words to the dictionary is also kind of sad. Using a dictionary within the existing apps such as the TextEditor, Browser and HackStudio is bit of a more compelling use case than a standalone dictionary application that's only slightly easier to open than a google search for a word definition.

**Comment by rogual** [comment]
OK. I thought maybe a basic dictionary would be a good step in the direction of a good dictionary.

I mean, I've enjoyed playing around with SerenityOS and — isn't that the case with all the apps? Like, isn't every app in the system currently just a good step in the direction of something more complete?

I suppose I've seen Andreas having fun hacking on things on stream, making something kinda neat as a start and saying, "this'll do for now" and thought it would be fun to contribute because I like working that way too, but what I'm hearing here is that code from newbies has to be very polished and immediately useful to make it in.

I've decided to cut my losses at this point and bow out. I do wish you the best of luck with the OS though, it does look really great.

**Pull request closed without merging**
